src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,355 @@
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualBasic;
+using static System.Runtime.InteropServices.JavaScript.JSType;
+
+#pragma warning disable CS8601 // Possible null reference assignment.
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+#pragma warning disable 8500 // address / sizeof of managed types
+
+namespace System.Numerics.Tensors
+{
+    public sealed class Tensor<T>
+        : IDisposable,
+          ITensor<Tensor<T>, T>
+        where T : IEquatable<T>
+    {
+        /// <summary>A byref or a native ptr.</summary>
+        internal readonly T[] _values;
+        /// <summary>The number of elements this Tensor contains.</summary>
+        internal readonly nint _linearLength;
+        /// <summary>The lengths of each dimension.</summary>
+        internal readonly nint[] _lengths;
+        internal readonly nint[] _strides;
+        /// <summary>If the backing memory is permanently pinned (so not just using a fixed statement).</summary>
+        internal readonly bool _isPinned;
+
+        /// <summary>
+        /// Creates a new empty Tensor.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor()
+        {
+            _linearLength = 0;
+            _values = [];
+            _lengths = [];
+            _strides = [];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor(T[] values, nint[] lengths, bool isPinned)
+        {
+            _linearLength = SpanHelpers.CalculateTotalLength(ref lengths);
+
+            Debug.Assert(_linearLength >= 0);
+
+            _values = values;
+            _lengths = lengths;
+            _strides = SpanHelpers.CalculateStrides(Rank, _lengths);
+            _isPinned = isPinned;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Tensor(T[] values, nint[] lengths, nint[] strides, bool isPinned)
+        {
+            _linearLength = SpanHelpers.CalculateTotalLength(ref lengths);
+
+            Debug.Assert(_linearLength >= 0);
+
+            _values = values;
+            _lengths = lengths;
+            _strides = strides;
+            if (strides == Array.Empty<nint>())
+                _strides = SpanHelpers.CalculateStrides(Rank, lengths);
+            _isPinned = isPinned;
+
+        }
+
+        // IDisposable
+
+        public void Dispose()
+        {
+        }
+
+        private static class EmptyTensor
+        {
+#pragma warning disable CA1825 // this is the implementation of Tensor.Empty<T>()
+            internal static readonly Tensor<T> Value = new Tensor<T>();
+#pragma warning restore CA1825
+        }
```

Comment:
Why is this nested type needed? i.e. why not just put this field on `Tensor<T>`?

Reply:
I just copied how it was done for [`System.Array`](src\libraries\System.Private.CoreLib\src\System\Array.cs), though I am happy to change it if you think that would be better.

Reply:
It's done like that on array so that we don't pay for an empty T[] for every T that's used with arrays, as arrays are ubiquitous and used everywhere. That's not the case with Tensor, which can just have `public static Tensor<T> Empty { get; } = new();`.

Reply:
KK. Fixed.